### PR TITLE
*: remove useless code 

### PR DESCRIPTION
**Pull request by mccxj**
1. func isConstraintKeyTp check `the first column of constraint is auto_increment`, code can be more clearly.
2. func XXXTableStmt check `Table == nil`, It looks like it's not necessary.
3. func checkAlterTableGrammar only check first new column, It is better to check multi new columns, though this feature is not supported at present.
4. fix typo.

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/5343)

### CODE REVIEW

**Comment by jackysp** [comment]
please add description for the pull request

**Comment by mccxj** [comment]
@jackysp @coocood  PTAL.

File: plan/preprocess.go [hunk 1]
```diff
@@ -139,13 +139,12 @@ func isConstraintKeyTp(constraints []*ast.Constraint, colDef *ast.ColumnDef) boo
...
 		// If the constraint as follows: primary key(c1, c2)
 		// we only support c1 column can be auto_increment.
-		if colDef.Name.Name.L != c.Keys[0].Column.Name.L {
-			continue
-		}
-		switch c.Tp {
-		case ast.ConstraintPrimaryKey, ast.ConstraintKey, ast.ConstraintIndex,
-			ast.ConstraintUniq, ast.ConstraintUniqIndex, ast.ConstraintUniqKey:
-			return true
+		if colDef.Name.Name.L == c.Keys[0].Column.Name.L {
```
Comment (coocood) [inline plan/preprocess.go]: I don't think we need to change here, it's a different style but not obviously better.

**Comment by jackysp** [comment]
@XuHuaiyu PTAL

File: plan/preprocess.go [hunk 2]
```diff
@@ -315,10 +300,12 @@ func (p *preprocessor) checkAlterTableGrammar(stmt *ast.AlterTableStmt) {
 				return
 			}
 		}
-		if len(spec.NewColumns) > 0 && spec.NewColumns[0] != nil {
-			if err := checkColumn(spec.NewColumns[0]); err != nil {
-				p.err = err
-				return
+		if len(spec.NewColumns) > 0 {
+			for _, colDef := range spec.NewColumns {
+				if err := checkColumn(colDef); err != nil {
```
Comment (XuHuaiyu) [inline plan/preprocess.go]: s/ err := / p.err :=

File: plan/preprocess.go [hunk 3]
```diff
@@ -315,10 +300,12 @@ func (p *preprocessor) checkAlterTableGrammar(stmt *ast.AlterTableStmt) {
 				return
 			}
 		}
-		if len(spec.NewColumns) > 0 && spec.NewColumns[0] != nil {
-			if err := checkColumn(spec.NewColumns[0]); err != nil {
-				p.err = err
-				return
+		if len(spec.NewColumns) > 0 {
```
Comment (coocood) [inline plan/preprocess.go]: This line is not needed.

File: expression/column.go [hunk 4]
```diff
@@ -189,8 +188,7 @@ func (col *Column) String() string {
 
 // MarshalJSON implements json.Marshaler interface.
 func (col *Column) MarshalJSON() ([]byte, error) {
-	buffer := bytes.NewBufferString(fmt.Sprintf("\"%s\"", col))
-	return buffer.Bytes(), nil
+	return []byte(fmt.Sprintf("\"%s\"", col)), nil
```
Comment (zz-jason) [inline expression/column.go]: why not just calling `col.String()` ?

File: plan/preprocess.go (hunk 1 above)
Comment (mccxj) [reply plan/preprocess.go]: OK.I will revert it.

File: expression/column.go (hunk 4 above)
Comment (mccxj) [reply expression/column.go]: Because there is double quote around it.

File: plan/preprocess.go (hunk 3 above)
Comment (mccxj) [reply plan/preprocess.go]: OK.

File: plan/preprocess.go (hunk 2 above)
Comment (mccxj) [reply plan/preprocess.go]: OK.

File: plan/preprocess.go [hunk 5]
```diff
@@ -348,7 +335,7 @@ func (p *preprocessor) checkAlterTableGrammar(stmt *ast.AlterTableStmt) {
 
 // checkDuplicateColumnName checks if index exists duplicated columns.
 func checkDuplicateColumnName(indexColNames []*ast.IndexColName) error {
-	for i := 0; i < len(indexColNames); i++ {
+	for i := 0; i < len(indexColNames)-1; i++ {
```
[the remark this answers is not in the document]
Comment (mccxj) [reply plan/preprocess.go]: The last loop is not need.

**Comment by coocood** [comment]
LGTM

File: expression/aggregation/aggregation.go [hunk 6]
```diff
@@ -184,21 +184,20 @@ func (af *aggFunction) Equal(b Aggregation, ctx context.Context) bool {
 
 // String implements fmt.Stringer interface.
 func (af *aggFunction) String() string {
-	result := af.name + "("
+	buffer := bytes.NewBufferString(fmt.Sprintf("%s(", af.name))
```
Comment (tiancaiamao) [inline expression/aggregation/aggregation.go]: As long as you use a string bytes.Buffer, why not avoid the `fmt.Sprintf`:

```
var buffer bytes.Buffer
fmt.Fprintf(&buffer, "%s(")
```

File: expression/scalar_function.go [hunk 7]
```diff
@@ -50,21 +50,20 @@ func (sf *ScalarFunction) GetCtx() context.Context {
 
 // String implements fmt.Stringer interface.
 func (sf *ScalarFunction) String() string {
-	result := sf.FuncName.L + "("
+	buffer := bytes.NewBufferString(fmt.Sprintf("%s(", sf.FuncName.L))
```
Comment (tiancaiamao) [inline expression/scalar_function.go]: ditto

File: plan/preprocess.go [hunk 8]
```diff
@@ -218,11 +218,6 @@ func (p *preprocessor) checkDropDatabaseGrammar(stmt *ast.DropDatabaseStmt) {
 }
 
 func (p *preprocessor) checkCreateTableGrammar(stmt *ast.CreateTableStmt) {
-	if stmt.Table == nil {
```
Comment (tiancaiamao) [inline plan/preprocess.go]: Any test cover this change?
Or are your sure `go test -cover` never run to this line?

File: plan/preprocess.go [hunk 9]
```diff
@@ -315,9 +301,8 @@ func (p *preprocessor) checkAlterTableGrammar(stmt *ast.AlterTableStmt) {
 				return
 			}
 		}
-		if len(spec.NewColumns) > 0 && spec.NewColumns[0] != nil {
```
Comment (tiancaiamao) [inline plan/preprocess.go]: `spec.NewColumns[0] != nil` removed?

**Comment by coocood** [comment]
@mccxj 
Any update?

**Comment by zhexuany** [comment]
@mccxj 
a friendly ping ;)

File: plan/preprocess.go (hunk 8 above)
Comment (zhexuany) [reply plan/preprocess.go]: @mccxj Please address this comment.

File: plan/preprocess.go (hunk 9 above)
Comment (zz-jason) [reply plan/preprocess.go]: @tiancaiamao Under what situations can `spec.NewColumns[0] == nil` happen?

File: plan/preprocess.go (hunk 8 above)
Comment (zz-jason) [reply plan/preprocess.go]: @mccxj Why remove this check ?

**Comment by zimulala** [comment]
@mccxj 
Any update?

**Comment by zz-jason** [comment]
@mccxj I noticed that you have removed the nil check in these functions: `checkCreateTableGrammar`, `checkDropTableGrammar`, `checkAlterTableGrammar`.

Take `checkCreateTableGrammar` as an example, we may faile the syntax check while parsing a `create table` sql statement without table name being provided like `create table (a bigint, b bigint)`, can you add some tests to check theses behaviors, thus we are sure that it's safe to remove theses nil checks because our parser never let these situations happen.

File: expression/scalar_function.go (hunk 7 above)
Comment (mccxj) [reply expression/scalar_function.go]: OK.

File: plan/preprocess.go (hunk 8 above)
Comment (mccxj) [reply plan/preprocess.go]: parser.y shows ast.CreateTableStmt's Table will not be null.

**Comment by mccxj** [comment]
more testcase added. @zz-jason PTAL.

File: plan/preprocess.go [hunk 10]
```diff
@@ -341,13 +326,13 @@ func (p *preprocessor) checkAlterTableGrammar(stmt *ast.AlterTableStmt) {
...
 // checkDuplicateColumnName checks if index exists duplicated columns.
 func checkDuplicateColumnName(indexColNames []*ast.IndexColName) error {
-	for i := 0; i < len(indexColNames); i++ {
-		name1 := indexColNames[i].Column.Name
-		for j := i + 1; j < len(indexColNames); j++ {
-			name2 := indexColNames[j].Column.Name
-			if name1.L == name2.L {
-				return infoschema.ErrColumnExists.GenByArgs(name2)
-			}
+	colNames := make(map[string]string, len(indexColNames))
```
Comment (zz-jason) [inline plan/preprocess.go]: a `map[string]struct{}` is enough, we don't need to save the values.

**Comment by zz-jason** [comment]
@tiancaiamao @zimulala PTAL

**Pull request merged**